components/supervisor/pkg/ports/ports.go

```diff
@@ -356,15 +362,41 @@ func (pm *Manager) nextState() map[uint32]*managedPort {
 }
 
 func (pm *Manager) autoExpose(ctx context.Context, mp *managedPort, public bool) {
-	err := pm.E.Expose(ctx, mp.LocalhostPort, mp.GlobalPort, public)
+	err := pm.autoExposeWithBackoff(ctx, mp, public)
 	if err != nil {
-		log.WithError(err).WithField("port", *mp).Warn("cannot auto-expose port")
+		if err != context.Canceled {
+			log.WithError(err).WithField("port", *mp).Warn("cannot auto-expose port")
+		}
 		return
 	}
 	pm.autoExposed[mp.LocalhostPort] = mp.GlobalPort
 	log.WithField("port", *mp).Info("auto-expose port")
 }
 
+func (pm *Manager) autoExposeWithBackoff(ctx context.Context, mp *managedPort, public bool) error {
+	delay := pm.minAutoExposeDelay
+	for {
+		err := pm.E.Expose(ctx, mp.LocalhostPort, mp.GlobalPort, public)
```

Comment:
Nit: I'd move the retry and exponential backoff into [`GitpodExposedPorts`](https://github.com/gitpod-io/gitpod/commit/a92a928c3f73c26937bd4df9600a25dde582eb34)